### PR TITLE
Best and Worst GCC Compiler Flags for Embedded

### PR DESCRIPTION
**Pull request by chrisc11**
A fun post about some of my most and least favorite compiler flags :)

### CODE REVIEW

File: _drafts/best-and-worst-gcc-clang-compiler-flags.md [hunk 1]
```diff
@@ -0,0 +1,851 @@
...
+stack_usage.c:14:5: warning: stack usage might be unbounded [-Wstack-usage=]
+ int vla_stack_usage(int num) {
+```
+
+## The Best Options for Debugability & Code Size
+
+### -g3
+
+When a binary is loaded on an embedded device, it's important debug information is included in the
+ELF. Otherwise, gdb (or your IDE) will not be able to display variables or backtraces. There's many
```
Comment (franc0is) [inline _drafts/best-and-worst-gcc-clang-compiler-flags.md]: There *are* many "Debug Option" flags

File: _drafts/best-and-worst-gcc-clang-compiler-flags.md [hunk 2]
```diff
@@ -0,0 +1,851 @@
...
+
+(gdb) p/x MEMFAULT_EXAMPLE_MACRO
+$1 = 0x4d
+```
+
+### -Os
+
+You can find all the details about optimization flags in the "Optimization Options" section of the
+GNU GCC docs[^5]. `-Os`, optimize for size, is generally the optimization flag you will see used
+for embedded systems. Its enables a good balance of flags which optimize for size _as well as
```
Comment (franc0is) [inline _drafts/best-and-worst-gcc-clang-compiler-flags.md]: s/Its/It enables

File: _drafts/best-and-worst-gcc-clang-compiler-flags.md [hunk 3]
```diff
@@ -0,0 +1,851 @@
...
+speed_. Forgetting to flip on this flag, can have serious code size impacts. For more details check
```
Comment (franc0is) [inline _drafts/best-and-worst-gcc-clang-compiler-flags.md]:
```
on this flag, can have serious 
            ^
 superfluous comma
```

File: _drafts/best-and-worst-gcc-clang-compiler-flags.md [hunk 4]
```diff
@@ -0,0 +1,851 @@
...
+By default the linker will place all functions in an object within the same linker "section". This
+becomes very clear when examining the mapfile and seeing a bunch of symbols in the `.text` section.
+
+```
+ .text          0x0000000000008238       0x60 /var/folders/dm/b0yt_1d11z9d53rtsm3y74sm0000gn/T//ccqZLzPH.o
+                0x0000000000008238                simple_math_get_sum
+                0x0000000000008268                simple_math_get_delta
+```
+
+With `-ffunction-sections`, each function gets it's own section:
```
Comment (franc0is) [inline _drafts/best-and-worst-gcc-clang-compiler-flags.md]: each function gets **its** own section

File: _drafts/best-and-worst-gcc-clang-compiler-flags.md [hunk 5]
```diff
@@ -0,0 +1,851 @@
...
+  one version is 255 and then in a future version it's 4096, the size of the enum will change from a
+  1-byte to a 2-byte representation. This means linking against a pre-compiled library using the older version
+  of the enum will no longer work.
+- When operations are performed on sizes that don't match the size of the architecture, extra
+  instructions are often needed to mask out bits in the register, which uses more code space. This
+  can easily be seen with the for loop example below.
+- Enum types can wind up in packed structures. Changing the legal values of an enum can then cause
+  unexpected breakages because the size of the enum member and consequently the structure will
+  change. If an enum is used in a struct saved on flash, for example, after a firmware update the
+  code may no longer be reading the data from flash correctly.
```
Comment (franc0is) [inline _drafts/best-and-worst-gcc-clang-compiler-flags.md]: may no longer be able to read the data from flash correctly

File: _drafts/best-and-worst-gcc-clang-compiler-flags.md [hunk 6]
```diff
@@ -0,0 +1,851 @@
...
+
+Per the documentation[^16], this flag "[breaks] IEEE or ISO rules/specifications for math function" in
+the name of faster code. It should be avoided!
+
+### -Wunused-parameter
+
+This warning will detect when a parameter passed to a function is never used in the
+function. While the goal here is to detect functions with dead parameters, as you start to build
+abstractions, sometimes parameters simply won't be used by all the implementations and the code
+needs to be littered marking parameters as unused with some sort of macro like
```
Comment (franc0is) [inline _drafts/best-and-worst-gcc-clang-compiler-flags.md]: the code needs to be cluttered by the tricks you will need to mark those parameters as used:

File: _drafts/best-and-worst-gcc-clang-compiler-flags.md [hunk 7]
```diff
@@ -0,0 +1,851 @@
...
+Compilers have hundreds of flags and configuration settings which can be toggled to control
+performance optimizations, code size, error checks and diagnostic information emitted. Often
+these settings wind up being copy and pasted from project to project, makefile to makefile, but
+it's good to periodically audit the current options selected for a project.
+
+<!-- excerpt start -->
+
+In this article we will explore some of the best and worst compiler flags for GCC (and Clang). Our
+focus will be on flags used for embedded projects but the reasoning applies to other development environments as well. We will explore the impact
+the flag has by walking through practical C code examples.
```
Comment (franc0is) [inline _drafts/best-and-worst-gcc-clang-compiler-flags.md]: We will explore the impact *each* flag has ..

**Comment by chrisc11** [comment]
Thanks for the read!

> I've caught a couple of small things. I'm sad about LTO being under "worst" flags tho.

Maybe someday it will get promoted! ;) There's definitely a lot of work going on in the area. I think for embedded some kind of way to only lto certain sub-units like a static library would be really neat!

File: _drafts/best-and-worst-gcc-clang-compiler-flags.md [hunk 8]
```diff
@@ -0,0 +1,892 @@
...
+
+Here's a c snippet that will happily compile but has a couple unfortunate bugs.
+
+```c
+static void prv_various_memset_bugs(void) {
+  #define NUM_ITEMS 10
+  uint32_t buf[NUM_ITEMS];
+  memset(buf, NUM_ITEMS, 0x0);
+  memset(buf, 0x0, NUM_ITEMS);
+  }
```
Comment (tyhoff) [inline _drafts/best-and-worst-gcc-clang-compiler-flags.md]: I believe most of your braces in your code examples should be "tabbed left" once.

File: _drafts/best-and-worst-gcc-clang-compiler-flags.md [hunk 9]
```diff
@@ -0,0 +1,892 @@
...
+    int result = i;
+    // do something with result
+  }
+
+  return result;
+}
+```
+
+```bash
+EXTRA_CFLAGS=-Wshadow make
```
Comment (tyhoff) [inline _drafts/best-and-worst-gcc-clang-compiler-flags.md]: When you show these shell commands, prefix every one of them with `$ `. The commands don't stick out super well.

File: _drafts/best-and-worst-gcc-clang-compiler-flags.md [hunk 10]
```diff
@@ -0,0 +1,892 @@
...
+etc) a set of implicit conversions on the types can take place. First **integer promotion** takes
+place. The C standard defines this as[^17]:
+
+> If an **int** can represent all values of the original type (as restricted by the width, for a
+> bit-field), the value is converted to an **int**; otherwise, it is converted to an **unsigned
+> int**. These are called the **integer promotions**.)
+
+Then the following set of rules are applied to the promoted values:
+
+> ![](img/compiler-flags/arithmetic-conversions-int.png)
```
Comment (tyhoff) [inline _drafts/best-and-worst-gcc-clang-compiler-flags.md]: This may be broken.

Comment (chrisc11) [reply _drafts/best-and-worst-gcc-clang-compiler-flags.md]: gah, wasn't committed, fixed!

File: _drafts/best-and-worst-gcc-clang-compiler-flags.md (hunk 8 above)
Comment (chrisc11) [reply _drafts/best-and-worst-gcc-clang-compiler-flags.md]: gah, auto-formatter keeps doing that, looks like I only missed correcting 2 :)

**Pull request merged**